### PR TITLE
feat: add storage package

### PR DESCRIPTION
**Pull request by verbanicm**
@dcreey needs this for iam-drift to create single shared package.

### CODE REVIEW

File: pkg/storage/google_cloud_storage.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	defer cancel()
+
+	attrs, err := o.Attrs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get object metadata: %w", err)
+	}
+	return attrs.Metadata, nil
+}
+
+func (s *GoogleCloudStorage) DeleteObject(ctx context.Context, bucket, name string, ignoreNotFound bool) error {
```
Comment (sethvargo) [inline pkg/storage/google_cloud_storage.go:167]: ignoreNotFound feels like a weird API; why would we _not_ want to ignore it? The caller says "ensure object FOO is deleted". If it doesn't exist, the invariant is satisfied.

File: pkg/storage/google_cloud_storage.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	defer cancel()
+
+	writer := o.NewWriter(ctx)
+
+	if contentType != "" {
+		writer.ContentType = contentType
+	}
+
+	if metadata != nil {
+		writer.Metadata = metadata
```
[the remark this answers is not in the document]
Comment (verbanicm) [reply pkg/storage/google_cloud_storage.go]: I feel like that was javascript specific, the docs mention this property is for user provided metadata, not all metadata. Made the change anyway.

File: pkg/storage/google_cloud_storage.go [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/googleapis/gax-go/v2"
+)
+
+const MiB = 1_048_576 // 1 MiB
```
Comment (sethvargo) [inline pkg/storage/google_cloud_storage.go:28]: Nit: `1 << 20` might be more explicit

File: pkg/storage/google_cloud_storage.go [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+func (s *GoogleCloudStorage) UploadObject(ctx context.Context, bucket, name string, contents []byte, opts ...UploadOption) error {
+	cfg := makeUploadConfig(len(contents), opts)
+
+	o, ctx, cancel := s.objectHandleWithRetries(ctx, bucket, name)
+	defer cancel()
+
+	writer := o.If(storage.Conditions{DoesNotExist: true}).NewWriter(ctx)
+	defer writer.Close()
```
Comment (sethvargo) [inline pkg/storage/google_cloud_storage.go:123]: You still want to check the error return value here though.

File: pkg/storage/google_cloud_storage.go [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+	}
+
+	if _, err := writer.Write(contents); err != nil {
+		return fmt.Errorf("failed to write data: %w", err)
+	}
+
+	return nil
+}
+
+func (s *GoogleCloudStorage) DownloadObject(ctx context.Context, bucket, name string) (io.ReadCloser, error) {
```
Comment (sethvargo) [inline pkg/storage/google_cloud_storage.go:149]: Doc: it's up to the caller to close the reader

Comment (verbanicm) [reply pkg/storage/google_cloud_storage.go:149]: Should this go in the interface doc or the implementation....or both?

**Pull request merged**